File: Component/BusiLib/src/main/java-dialog/com/zq/dialog/adapter/ReportAdapter.java
package com.zq.dialog.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;

import com.common.view.recyclerview.DiffAdapter;
import com.component.busilib.R;
import com.zq.dialog.model.ReportModel;

public class ReportAdapter extends DiffAdapter<ReportModel, RecyclerView.ViewHolder> {

    RecyclerOnItemCheckListener onItemCheckListener;

    public ReportAdapter(RecyclerOnItemCheckListener onItemCheckListener) {
        this.onItemCheckListener = onItemCheckListener;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.report_item_layout, parent, false);
        ReportItemHolder viewHolder = new ReportItemHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ReportModel model = mDataList.get(position);

        ReportItemHolder reportItemHolder = (ReportItemHolder) holder;
        reportItemHolder.bind(model);

    }

    @Override
    public int getItemCount() {
        return mDataList.size();
    }

    private class ReportItemHolder extends RecyclerView.ViewHolder {

        CheckBox mCheckBox;
        ReportModel mReportModel;

        public ReportItemHolder(View itemView) {
            super(itemView);
            mCheckBox = itemView.findViewById(R.id.checkbox);

            mCheckBox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton compoundButton, boolean isCheck) {
                    if (onItemCheckListener != null) {
                        onItemCheckListener.onCheckedChanged(isCheck, mReportModel);
                    }
                }
            });
        }

        public void bind(ReportModel model) {
            this.mReportModel = model;
            mCheckBox.setText(model.getText());
        }
    }

    public interface RecyclerOnItemCheckListener {
        void onCheckedChanged(boolean isCheck, ReportModel model);
    }
}

File: Component/BusiLib/src/main/java-report/com/component/report/adapter/ReportAdapter.kt
package com.component.report.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.common.view.DebounceViewClickListener
import com.common.view.ex.ExImageView
import com.component.busilib.R

class ReportAdapter : RecyclerView.Adapter<ReportAdapter.FeedReportViewHolder>() {

    var mDataList = ArrayList<ReportModel>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FeedReportViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.report_item_layout, parent, false)
        return FeedReportViewHolder(view)
    }

    override fun onBindViewHolder(holder: FeedReportViewHolder, position: Int) {
    }

    override fun onBindViewHolder(holder: FeedReportViewHolder, position: Int, payloads: MutableList<Any>) {
        super.onBindViewHolder(holder, position, payloads)
        if (payloads.isEmpty()) {
            holder.bindData(position, mDataList[position])
        } else {
            holder.refreshSelected(position, mDataList[position])
        }
    }

    override fun getItemCount(): Int {
        return mDataList.size
    }

    // 获得所有选中的类型
    fun getSelectedList(): ArrayList<Int> {
        var result = ArrayList<Int>()
        for (model in mDataList) {
            if (model.isSelected) {
                result.add(model.type)
            }
        }
        return result
    }

    inner class FeedReportViewHolder(var item: View) : RecyclerView.ViewHolder(item) {

        var mPosition = 0
        var mModel: ReportModel? = null

        val mContentTv: TextView = item.findViewById(R.id.content_tv)
        val mSelectIv: ExImageView = item.findViewById(R.id.select_iv)

        init {
            item.setOnClickListener(object : DebounceViewClickListener() {
                override fun clickValid(v: View?) {
                    clickItem()
                }
            })

            mSelectIv.setOnClickListener(object : DebounceViewClickListener() {
                override fun clickValid(v: View?) {
                    clickItem()
                }
            })
        }

        fun bindData(position: Int, model: ReportModel) {
            this.mPosition = position
            this.mModel = model

            mContentTv.text = model.reportDesc
            refreshSelected(position, model)
        }

        fun refreshSelected(position: Int, model: ReportModel) {
            this.mPosition = position
            this.mModel = model

            mSelectIv.isSelected = model.isSelected
        }

        fun clickItem() {
            mModel?.let {
                it.isSelected = !it.isSelected
                mDataList[mPosition] = it
                this@ReportAdapter.notifyItemChanged(mPosition, 0)
            }
        }
    }
}

